Introduce sound related system calls

ChinYikMing opened the pull request:
`mus2midi` utility is referenced by [Mus2Midi](https://github.com/fawtytoo/Mus2Midi) . 

Tested in Ubuntu 20.04.

Close #152

jserv commented on Makefile:
```diff
@@ -96,6 +101,7 @@ OBJS := \
 	decode.o \
 	io.o \
 	syscall.o \
+	mus2midi.o \
```
Append `mus2midi.o` to `OBJS_EXT` instead. That is,
```
OBJS_EXT += mus2midi.o syscall_sdl.o
```

jserv commented on src/syscall_sdl.c:
```diff
@@ -276,3 +337,162 @@ void syscall_submit_queue(riscv_t *rv)
+    Mix_CloseAudio();
+    Mix_Quit();
+
+    if (mid_data)
```
Simply rewrite as following:
```c
free(mid_data);
```

Quote from [free(3)](https://man7.org/linux/man-pages/man3/free.3p.html):
> If ptr is a null pointer, no action shall occur.

jserv commented on src/syscall_sdl.c:
```diff
@@ -276,3 +337,162 @@ void syscall_submit_queue(riscv_t *rv)
+    uint8_t m_data[MUSIC_MAX_SIZE];
+    memory_read(s->mem, m_data, m_data_offset, m_data_size);
+
+    sound_t music;
```
Use C99 [designated initializers](https://gcc.gnu.org/onlinedocs/gcc/Designated-Inits.html) instead.

jserv commented on src/mus2midi.h:
```diff
@@ -0,0 +1,13 @@
+ * "LICENSE" for information on usage and redistribution of this file.
+ */
+
+#ifndef __MUS2MIDI_H__
```
It is not worthy to have a header containing only one function prototype. Declare in `syscall_sdl.c` directly.

jserv commented on src/mus2midi.h:
```diff
@@ -0,0 +1,13 @@
+#ifndef __MUS2MIDI_H__
+#define __MUS2MIDI_H__
+
+typedef unsigned char BYTE;
```
Avoid pre-C99 stuff such as `BYTE`. Always use types such as `uint8_t`.

jserv commented on src/syscall_sdl.c:
```diff
@@ -276,3 +337,162 @@ void syscall_submit_queue(riscv_t *rv)
     }
 }
+
+void *sfx_handler(void *arg)
```
Add `static` qualifier.

jserv commented on src/syscall_sdl.c:
```diff
@@ -276,3 +337,162 @@ void syscall_submit_queue(riscv_t *rv)
+    return NULL;
+}
+
+void *music_handler(void *arg)
```
Ditto. `static` qualifer.

jserv commented on src/mus2midi.c:
```diff
@@ -0,0 +1,206 @@
+
+#include "mus2midi.h"
+
+typedef unsigned short WORD;
```
Utilize the types defined in `<stdint.h>`.

jserv commented on src/mus2midi.c:
```diff
@@ -0,0 +1,206 @@
+    char id[4];
+    WORD scoreLen;
+    WORD scoreStart;
+} __attribute__((packed)) HDR_MUS;
```
Check [CONTRIBUTING.md](https://github.com/sysprog21/rv32emu/blob/master/CONTRIBUTING.md):
> Use [snakecase](https://en.wikipedia.org/wiki/Snake_case). Do not use "camelcase".

jserv commented on src/mus2midi.c:
```diff
@@ -0,0 +1,206 @@
+BYTE magicEOT[4] = {0x00, 0xff, 0x2f, 0x00};
+
+int controllerMap[16] = {-1, 0,  1,   7,   10,  11,  91,  93,
+                         64, 67, 120, 123, 126, 127, 121, -1};
```
Add an extra comma before `}`, so that clang-format can behave better. i.e.,
```
    ..., 27, 121, -1,
};
```

jserv commented on src/mus2midi.c:
```diff
@@ -0,0 +1,206 @@
+    WORD ticks;
+} __attribute__((packed)) HDR_MID;
+
+char magicMus[4] = {'M', 'U', 'S', 0x1a};
```
Hide the internal symbols as possible as you can. That is, add `static` qualifers.

jserv commented on src/mus2midi.c:
```diff
@@ -0,0 +1,206 @@
+BYTE deltaBytes[4];
+int deltaCount;
+
+// maintain a list of channel volume
```
Use C-style comments.

jserv reviewed (changes requested):
Update `docs/syscall.md` and provide a test program in `tests` directory.

ChinYikMing commented:
> Provide a test program in `tests` directory.

Sound version of `doom.elf` is come with the commit, simply run `make doom` in `sound` branch

jserv commented on src/syscall.c:
```diff
@@ -18,21 +18,27 @@
+    _(clock_gettime,        403)     \
+    _(open,                 1024)    \
+    IIF(RV32_HAS(SDL))(              \
+        _(play_sfx,         0xBABE)  \
```
You proposed 6 system calls for audio/sound. i.e.,
* play_sfx
* play_music
* stop_music
* set_music_volume
* setup_sound
* shutdown_sound

Can you use fewer ones? We may discuss on #152 before code reviewing.

jserv commented on src/syscall_sdl.c:
```diff
@@ -276,3 +337,162 @@ void syscall_submit_queue(riscv_t *rv)
+     * hack of offset as DOOM is 32-bit but
+     * Emulator is 64-bit when accessing the musicinfo_t
+     */
+    uint32_t m_data_offset = *(uint32_t *) ((uint8_t *) &m + 8);
```
The prefix `m_` is useless. Older compilers and development environments were stupid, and old programmers manually made use of such hacky prefix. Now, drop them.

jserv commented:
> Sound version of `doom.elf` is come with the commit, simply run `make doom` in `sound` branch

No, let's minimize the necessary changes. You should provide a minimal C program that demonstrates the use of the proposed system calls instead of making extensive modifications to Doom. Once this pull request is merged, we can proceed with the Doom and Quake pull requests in their respective repositories.

jserv commented:
[Simple SDL2 Audio](https://github.com/jakebesworth/Simple-SDL2-Audio) operates independently of SDL2 Mixer and is designed to play a single music file at a time. It also supports an unlimited number of sounds, achieving audio mixing natively without the need for an external mixer. It might be useful.

ChinYikMing commented:
> Update `docs/syscall.md` and provide a test program in `tests` directory.

You may test the result of integration using the test program in `tests/sound` directory, use `=h` option to check the usage. 

To test all at once, simply run `../../build/rv32emu sound.elf =m d_e1m1.mus =s dspistol =upmv =srep` in `tests/sound` directory.

jserv commented on README.md:
```diff
@@ -35,9 +35,10 @@ Features:
-* macOS: `brew install sdl2`
-* Ubuntu Linux / Debian: `sudo apt install libsdl2-dev`
+To ensure proper operation, the target system should have the [SDL2 library](https://www.libsdl.org/) 
+and [SDL2_Mixer library](https://wiki.libsdl.org/SDL2_mixer/FrontPage) installed.
```
Use shorter URL. i.e., `https://wiki.libsdl.org/SDL2_mixer`

jserv reviewed (changes requested):
Rebase the latest `master` branch.

jserv commented on docs/syscall.md:
```diff
@@ -95,9 +112,9 @@ Currently only a couple system calls are supported via semihosting, which enable
+## Experimental Display, Event, and Sound System Calls
 
-These system calls are solely for the convenience of accessing the [SDL library](https://www.libsdl.org/) and are only intended for the presentation of RISC-V graphics applications. They are not present in the ABI interface of POSIX or Linux.
+These system calls are solely for the convenience of accessing the [SDL library](https://www.libsdl.org/) and [SDL2_Mixer](https://wiki.libsdl.org/SDL2_mixer/FrontPage) and are only intended for the presentation of RISC-V graphics applications. They are not present in the ABI interface of POSIX or Linux.
```
Ditto. Always use shorter URL.

jserv commented on docs/syscall.md:
```diff
@@ -138,3 +155,31 @@ To inform the emulator that a batch of submissions should be processed, the appl
 
 * `RELATIVE_MODE_SUBMISSION`: Enable or disable the mouse relative mode. If the mouse relative mode is enabled, the mouse cursor is wrapped within the window border, it's associated with an 8-bit wide boolean value that indicates whether the relative mouse mode should be enbled.
+
+### `control_sound` - control the behavior of music and sound effect(sfx)
```
Follow the consistent manner described above.

jserv replied:
Don't mix sound and audio. Instead, distinguish them properly.

jserv commented on src/mus2midi.c:
```diff
@@ -0,0 +1,224 @@
+    char id[4];
+    uint16_t score_len;
+    uint16_t score_start;
+} __attribute__((packed)) hdr_mus_t;
```
Why not `music_header_t`?

ChinYikMing replied:
`mus` is the type of file format before conversion so I think stick with `mus_header_t` is OK

jserv commented on src/mus2midi.c:
```diff
@@ -0,0 +1,224 @@
+    uint16_t type;
+    uint16_t ntracks;
+    uint16_t ticks;
+} __attribute__((packed)) hdr_mid_t;
```
Why no `midi_header_t`?

jserv commented on src/mus2midi.c:
```diff
@@ -0,0 +1,224 @@
+    uint16_t ticks;
+} __attribute__((packed)) hdr_mid_t;
+
+static char magic_mus[4] = {
```
It is a bit confusing to specify abbreviations such as `mus`, `mid`, `trk`, etc. Use the fully qualified ones.

ChinYikMing replied:
`mus` is the type of file format before conversion so I think stick with `mus` is OK but I will rename `mid` to `midi` and `trk` to `track` instead

jserv replied:
Enforce `static const` qualifiers to objects with prefix `magic_`.

jserv commented on src/mus2midi.c:
```diff
@@ -0,0 +1,224 @@
+    -1, 0, 1, 7, 10, 11, 91, 93, 64, 67, 120, 123, 126, 127, 121, -1,
+};
+
+uint8_t *_mid_data;
```
Declare the internal symbols as `static`.

jserv commented on src/syscall.c:
```diff
@@ -18,21 +18,23 @@
+    _(clock_gettime,        403)     \
+    _(open,                 1024)    \
+    IIF(RV32_HAS(SDL))(              \
+        _(control_sound,    0xBABE)  \
```
Move the sound specific ones to the end.

jserv commented on docs/syscall.md:
```diff
@@ -155,3 +155,39 @@ To inform the emulator that a batch of submissions should be processed, the appl
 
 * `RELATIVE_MODE_SUBMISSION`: Enable or disable the mouse relative mode. If the mouse relative mode is enabled, the mouse cursor is wrapped within the window border, it's associated with an 8-bit wide boolean value that indicates whether the relative mouse mode should be enbled.
+
+### `control_sound` - control the behavior of music
```
The description is confusing. Don't mix the term sound and music.

jserv commented on src/mus2midi.c:
```diff
@@ -0,0 +1,224 @@
+/* maintain a list of channel volume */
+static uint8_t mus_channel[16];
+
+static void convert()
```
Rewrite as `static void convert(void)`.

jserv reviewed (changes requested):
Move the content of `src/mus2midi.c` to `src/syscall_sdl.c` since the former is short and dedicated to the latter.

jserv commented on docs/syscall.md:
```diff
@@ -155,3 +155,39 @@ To inform the emulator that a batch of submissions should be processed, the appl
+* `SET_MUSIC_VOLUME`: Update the music volume at some point if needed.
+
+#### Music
+Music data is defined in a structure called `musicinfo_t` which is from Doom application. The main data in the structure is `data` and `size`. Emulator use `data` and `size` to play music. Other fields are not used by now so the structure should be more generic to support other applications.
```
What about Quake porting?

ChinYikMing replied:
I introduce an ABI such that first two fields of structure  `musicinfo_t` and `sfxinfo_t` in any applications must be sound data and follows the sound data size, such that the sound system of emulator can compatible to any applications including Quake. A simple example is `tests/sound/sounds.c`.

 i.e.,
```c
typedef stuct {
    void *data;
    int size;
    ...
} musicinfo_t;
```

Simply ignore other fields of the structure for now.

jserv commented on docs/syscall.md:
```diff
@@ -155,3 +155,39 @@ To inform the emulator that a batch of submissions should be processed, the appl
+* `PLAY_SFX`: Play the sfx, note that too many sfx at the same time could cause channel to run out and the sfx cannot be played. The sfx is played by a new thread and the reason is same as `PLAY_MUSIC`
+
+#### Sound Effect(sfx)
+Sfx data is defined in a structure called `sfxinfo_t` which is also from Doom application. The main data in the structure is `data`, `size`, and `volume`. Emulator use `data`, `size`, and `volume` to play sfx. Other fields are not used by now so the structure should be more generic to support other applications.
```
Improve the English writing.

jserv replied:
[QuillBot](https://quillbot.com/) is your friend.

jserv commented on src/syscall.c:
```diff
@@ -18,21 +18,23 @@
+    _(clock_gettime,        403)     \
+    _(open,                 1024)    \
+    IIF(RV32_HAS(SDL))(              \
+        _(setup_audio,      0xD00D)  \
```
Again, don't mix audio and sound in mnemonic identifiers. `setup_audio` and `control_audio` should appear at the end of syscall listing.

ChinYikMing commented:
> Move the content of `src/mus2midi.c` to `src/syscall_sdl.c` since the former is short and dedicated to the latter.

tests/sound/sound.c requires `src/mus2midi.c`, I think we should leave it to be.

jserv commented on src/syscall.c:
```diff
@@ -18,21 +18,23 @@
+        _(draw_frame,       0xBEEF)  \
+        _(setup_queue,      0xC0DE)  \
+        _(submit_queue,     0xFEED)  \
+        _(setup_audio,      0xD00D)  \
```
We may swap the magic numbers.
The term 'dood' is an informal and playful variation of the word "dude," which is a slang term that originated in American English. "Dude" is often used to refer to a person, typically a man or a friend.

The setup of audio is called "baby," and then the control of audio is called "dood," which sounds reasonable.

jserv commented on tests/sound/sound.c:
```diff
@@ -0,0 +1,278 @@
+    int size;
+} musicinfo_t;
+
+typedef struct sfxinfo_struct {
```
Simply remove `sfxinfo_struct` as an anonymous structure.

jserv replied:
To avoid code duplication, let's move the structure definitions and the function prototype to the header, `src/mus2midi.h`.

jserv commented on src/mus2midi.c:
```diff
@@ -0,0 +1,223 @@
+ */
+
+/*
+ * Reference: https://github.com/fawtytoo/Mus2Midi
```
Add more comments. Take the example below:
> This is a basic MUS to MIDI converter designed for applications such as DOOM.

jserv replied:
Of course, it would be better to mention Quake adoption.

jserv commented on src/mus2midi.c:
```diff
@@ -0,0 +1,223 @@
+ */
+
+/*
+ * Reference: https://github.com/fawtytoo/Mus2Midi
```
Remove the hyperlink of Mus2Midi and put the license notice below:
```c
/* Copyright (C) 2021-2022 Steve Clark
 * 
 * This software is provided 'as-is', without any express or implied
 * warranty.  In no event will the authors be held liable for any damages
 * arising from the use of this software.
 *
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute it
 * freely, subject to the following restrictions:
 *
 * 1. The origin of this software must not be misrepresented; you must not
 *  claim that you wrote the original software. If you use this software
 *  in a product, an acknowledgment in the product documentation would 
 *  be
 *  appreciated but is not required. 
 * 2. Altered source versions must be plainly marked as such, and must not be
 *  misrepresented as being the original software.
 * 3. This notice may not be removed or altered from any source distribution.
*/
```

jserv replied:
See https://github.com/fawtytoo/Mus2Midi/blob/main/Licence

jserv replied:
`rv32emu` is licensed under an MIT-style license, which is compatible with the zlib license used by the `MusMidi` project. However, since the license specifies that "This notice may not be removed or altered from any source distribution," we are obligated to comply.

jserv commented on src/mus2midi.c:
```diff
@@ -0,0 +1,223 @@
+/* maintain a list of channel volume */
+static uint8_t mus_channel[16];
+
+static void convert(void)
```
Attempt to add comments.

jserv commented on src/mus2midi.c:
```diff
@@ -0,0 +1,223 @@
+    midi_header_t midi_hdr;
+    uint8_t *mid_track_len;
+    int track_len;
+    int i;
```
Reduce the scope of `i`. i.e., move it into the `for` iteration below.

jserv commented on src/mus2midi.c:
```diff
@@ -0,0 +1,223 @@
+    int track_len;
+    int i;
+
+    if (strncmp(mus_hdr->id, magic_mus, 4) != 0)
```
Rewrite as following:
```c
if (!strncmp(mus_hdr->id, magic_mus, 4))
```

jserv commented on src/syscall_sdl.c:
```diff
@@ -7,13 +7,76 @@
+ * but normally only head in multiplayer
+ * since we are single player, stick with 11025 is OK
+ *
+ * In QUAKE, most sound effects with sample rate 11025
```
Always refer it as [Quake](https://en.wikipedia.org/wiki/Quake_(video_game)).

jserv commented on src/syscall_sdl.c:
```diff
@@ -276,3 +339,222 @@ void syscall_submit_queue(riscv_t *rv)
+    sound_t *music = (sound_t *) arg;
+    int looping = music->looping ? -1 : 1;
+
+    if (mid_data)
```
Simply rewrite as following:
```c
    free(mid_data).
```

jserv commented on src/syscall_sdl.c:
```diff
@@ -7,13 +7,76 @@
+
+/* SDL-mixer-related and music-related variables */
+static pthread_t music_thr;
+static uint8_t *mid_data;
```
Rename it to `midi_data`. The original term `mid_data` might lead to confusion, as it could be interpreted as either "middle data" or "data in the middle."

ChinYikMing replied:
I rename it as `music_midi_data` to avoid naming collision with `midi_data` that used by `mus2mid` converter since I have moved the `src/mus2midi.c` to `src/syscall_sdl.c`

jserv commented on src/syscall_sdl.c:
```diff
@@ -7,13 +7,76 @@
+} sound_t;
+
+/* SDL-mixer-related and music-related variables */
+static pthread_t music_thr;
```
Rename it to `music_thread`.

jserv commented on src/syscall_sdl.c:
```diff
@@ -7,13 +7,76 @@
+static Mix_Music *mid;
+
+/* SDL-mixer-related and sfx-related variables */
+static pthread_t sfx_thr;
```
Ditto. Use formal term "thread" rather than `thr`.

jserv commented on src/syscall_sdl.c:
```diff
@@ -7,13 +7,76 @@
+static pthread_t sfx_thr;
+static Mix_Chunk *sfx_chunk;
+static uint8_t *sfx_samples;
+static uint32_t sfx_sample_nums;
```
Rename to `nr_sfx_samples`, following Linux kernel naming scheme.
See https://stackoverflow.com/questions/8306202/what-does-nr-stand-for-in-system-call-number-that-is-usually-used-as-suffix

jserv commented on src/syscall_sdl.c:
```diff
@@ -276,3 +339,222 @@ void syscall_submit_queue(riscv_t *rv)
+    case INIT_AUDIO:
+        init_audio();
+        break;
+    case SHUTDOWN_AUDIO:
```
I am unsure whether it is meaningful to offer operations like initialization and shutdown for users, considering that the emulator manages the actual lifecycle of the audio instance. Consider the framebuffer management system call, `draw_frame`: the existence of a single system call is due to the fact that we only enable users to update the available screen, without providing functionality for initializing or closing the window.

ChinYikMing replied:
Most games like Doom and Quake, they will maintain their own sound system or server so there should be a pair of initialization and destroy semantic. In order to keep the semantic clear, I think we should provide operations like initialization and shutdown for users. Otherwise, the emulator will keep checking if the audio is enabled or not when to play the music or sound effects.

jserv replied:
Document the above considerations in `docs/` directory.

jserv commented on tests/sound/sound.c:
```diff
@@ -0,0 +1,278 @@
+    sfx->size = 5669; /* FIXME: hardcoded size */
+
+    FILE *_sfx = fopen(sfx_src, "rb");
+    assert(_sfx != NULL);
```
Rewrite as `assert(_sfx)`.

jserv commented on tests/sound/sound.c:
```diff
@@ -0,0 +1,278 @@
+    assert(sfx->data != NULL);
+    sfx->size = 5669; /* FIXME: hardcoded size */
+
+    FILE *_sfx = fopen(sfx_src, "rb");
```
Avoid the naming scheme starting with underscore.

Here is what the C standard states (section 7.1.3):
* All identifiers that start with an underscore followed by an uppercase letter or another underscore are always reserved for any purpose.
* All identifiers that start with an underscore are always reserved to be used as identifiers with file scope in both the ordinary and tag name spaces.

(The section proceeds to enumerate specific identifiers and groups of identifiers that are reserved by certain standard headers.)

The implication is that, for example, the implementation (whether it's the compiler or a standard header) can utilize the name `__FOO` for any purpose it deems fit.

In light of this, it's best to abstain from defining any such identifiers in your own code (unless your code is part of a C implementation -- and if you are uncertain, it likely is not). There's no real necessity to define these identifiers anyway; there is an abundance of unreserved identifiers at your disposal.

You can employ an identifier like `_foo` as long as it is confined to a local definition (not at file scope). However, I personally find it much simpler to avoid the use of leading underscores entirely.

jserv commented on tests/sound/sound.c:
```diff
@@ -0,0 +1,278 @@
+static void load_music()
+{
+    music = malloc(sizeof(musicinfo_t));
+    assert(music != NULL);
```
Ditto. Always write shorter and simpler code.

jserv commented on Makefile:
```diff
@@ -53,12 +53,17 @@ ifeq (, $(shell which sdl2-config))
+OBJS_EXT += mus2midi.o syscall_sdl.o
 $(OUT)/syscall_sdl.o: CFLAGS += $(shell sdl2-config --cflags)
 LDFLAGS += $(shell sdl2-config --libs)
+LDFLAGS += -lSDL2_mixer
```
Rewrite as following:
```
LDFLAGS += $(shell pkg-config --libs SDL2_mixer)
```

jserv commented on Makefile:
```diff
@@ -53,12 +53,17 @@ ifeq (, $(shell which sdl2-config))
 $(warning No sdl2-config in $$PATH. Check SDL2 installation in advance)
 override ENABLE_SDL := 0
 endif
+ifeq (0, $(shell pkg-config --exists SDL2_mixer && echo 1 || echo 0))
```
Replace this line with the following:
```
ifeq (1, $(shell pkg-config --exists SDL2_mixer; echo $$?))
```

ChinYikMing commented:
> > Move the content of `src/mus2midi.c` to `src/syscall_sdl.c` since the former is short and dedicated to the latter.
> 
> tests/sound/sound.c requires `src/mus2midi.c`, I think we should leave it to be.

I made  a mistake which `mus2midi` is only used in emulator since it control the sound, I will adopt the required change.

jserv commented on src/syscall_sdl.c:
```diff
@@ -7,13 +7,311 @@
 
 #include "state.h"
 
+/* Copyright (C) 2021-2022 Steve Clark
```
s/Copyright/Portions Copyright/

jserv replied:
Rearrange these newly-added lines immediately preceding the function `sfx_handler` for enhanced readability.

jserv commented on src/syscall_sdl.c:
```diff
@@ -7,13 +7,311 @@
+ *
+ * 1. The origin of this software must not be misrepresented; you must not
+ *  claim that you wrote the original software. If you use this software
+ *  in a product, an acknowledgment in the product documentation would
```
Indent the notice.

jserv reviewed (changes requested):
Improve git commit messages.

jserv commented on src/syscall_sdl.c:
```diff
@@ -276,3 +337,457 @@ void syscall_submit_queue(riscv_t *rv)
     }
 }
+
+/* Copyright (C) 2021-2022 Steve Clark
```
Replace "Copyright" with "Portions Copyright"

jserv commented on src/syscall_sdl.c:
```diff
@@ -276,3 +337,457 @@ void syscall_submit_queue(riscv_t *rv)
+/*
+ * This is a basic MUS to MIDI converter designed for applications such as DOOM.
+ *
+ * For Quake, it only have sound effects which are in WAV format so sfx_handler
```
Improve the English writing.

jserv commented on docs/syscall.md:
```diff
@@ -155,3 +155,41 @@ To inform the emulator that a batch of submissions should be processed, the appl
 
 * `RELATIVE_MODE_SUBMISSION`: Enable or disable the mouse relative mode. If the mouse relative mode is enabled, the mouse cursor is wrapped within the window border, it's associated with an 8-bit wide boolean value that indicates whether the relative mouse mode should be enbled.
+
+### `control_audio` - control the behavior of music
```
Refine the following documentation.

ChinYikMing replied:
My bad, I have swap `D00D` and `BABE`

jserv commented on src/syscall_sdl.c:
```diff
@@ -276,3 +337,458 @@ void syscall_submit_queue(riscv_t *rv)
+    char id[4];
+    uint16_t score_len;
+    uint16_t score_start;
+} __attribute__((packed)) mus_header_t;
```
Wrap `__attribute__((packed))` via the macro defined in `src/common.h. Consider Win32/MSVC compatibility.

jserv commented on docs/syscall.md:
```diff
@@ -155,3 +155,41 @@ To inform the emulator that a batch of submissions should be processed, the appl
+
+### `control_audio` - control the behavior of music
+
+**system call number**: `0xBABE`
```
Check the magic number.

jserv commented on docs/syscall.md:
```diff
@@ -155,3 +155,41 @@ To inform the emulator that a batch of submissions should be processed, the appl
+#### Music
+Music data is defined in a structure called `musicinfo_t`. The SDL2_mixer library is used by the emulator to play music using the fields `data` and `size` in the structure.
+
+### `control_audio` - control the behavior of sound effect(sfx)
```
Merge the paragraph into the previous one.

jserv commented on src/syscall_sdl.c:
```diff
@@ -276,3 +337,460 @@ void syscall_submit_queue(riscv_t *rv)
+ * format.
+ */
+
+__PACKED(typedef struct {
```
Rewrite as following:
```c
typedef __PACKED(struct {
    char id[4];
    uint16_t score_len;
    uint16_t score_start;
}) mus_header_t;
```

ChinYikMing replied:
I did this before but I think it will not work in MSVC

jserv replied:
Rework with macro `PACKED_STRUCT` mentioned earlier.

ChinYikMing replied:
My mistake, the rewrite suggestion previously is the one really work.

```c
struct x {
  char;
  double;
  short;
  char;
};
```
print result of  offset of the structure `x` compiled using MSVC:

no_pack_struct:
char = 0
double = 8
short = 16
char = 18

pack_struct:
char = 0
double = 1
short = 9
char = 11

jserv commented on src/syscall_sdl.c:
```diff
@@ -276,3 +337,460 @@ void syscall_submit_queue(riscv_t *rv)
+    /* sfx samples buffer */
+    sfx_samples = malloc(SFX_SAMPLE_SIZE);
+
+    // Initialize SDL2 Mixer
```
Enforce C-style comments.

jserv commented on tests/sound/sound.c:
```diff
@@ -0,0 +1,278 @@
+
+static void do_play_sound()
+{
+    if (has_music && has_sfx) {
```
Don't do this! Employ bitwise tricks. See https://github.com/jserv/ir/blob/dev/tools/ir.c
e.g.,
```c
    if (dump & IR_DUMP_DUMP) {
        ir_dump(ctx, f);
    }
```

jserv commented on src/common.h:
```diff
@@ -24,10 +24,30 @@
 #endif
 
+/* Packed macro */
+#if defined(__GNUC__) || defined(__clang__)
```
Rewrite as following:
```c
#if defined(__GNUC__) || defined(__clang__)
#define PACKED_STRUCT(name) struct __attribute__((packed)) name
#elif defined(_MSC_VER)
#define PACKED_STRUCT(name) \
    __pragma(pack(push, 1)) struct name __pragma(pack(pop))
#endif
```

Then this can be used like so:
```c
typedef PACKED_STRUCT(/*anonymous*/) { short a; int b } my_struct_t;
```

jserv commented on tests/sound/sound.c:
```diff
@@ -0,0 +1,281 @@
+static int music_volume = 3;
+static int sfx_volume = 15;
+
+#define MUSIC 1
```
Use `enum` instead of macros.

jserv commented on tests/sound/sound.c:
```diff
@@ -0,0 +1,281 @@
+
+static void do_play_sound()
+{
+    if (flag & MUSIC && flag & SFX) {
```
Can you rework as following?
```c
    if (flag & MUSIC)
        load_music();
    if (flag & SFX)
        load_sfx();
    ...

```

ChinYikMing committed:
Introduce sound related system calls

Incorporate the SDL2_mixer library to play sound in games
like Doom and Quake. The test program "tests/sound/sound.c"
illustrates the integration result. For more technical information,
see docs/syscall.md.

Close #152

jserv commented:
Thank @ChinYikMing for contributing!
